src/vsr/grid.zig

```diff
@@ -266,45 +270,91 @@ pub fn GridType(comptime Storage: type) type {
-        /// 1. Writes the free set to disk.
-        /// 2. Awaits all pending repair-writes to blocks that were just freed. This guarantees that
-        ///    there are no outstanding writes to (now-)free blocks when we enter the new checkpoint.
-        /// 3. Marks blocks released during the current checkpoint as free and eligible for
-        ///    acquisition in the next checkpoint.
-        /// 4. Marks the free set's own blocks as released (but not yet free).
+        pub fn open(grid: *Grid, callback: *const fn (*Grid) void) void {
+            assert(grid.callback == .none);
+
+            grid.callback = .{ .open = callback };
+            grid.free_set_checkpoint.open(
+                grid,
+                grid.superblock.working.free_set_reference(),
+                free_set_open_callback,
+            );
+        }
+
+        fn free_set_open_callback(free_set_checkpoint: *CheckpointTrailer) void {
+            const grid = @fieldParentPtr(Grid, "free_set_checkpoint", free_set_checkpoint);
+            const callback = grid.callback.open;
+
+            {
+                assert(!grid.free_set.opened);
+                defer assert(grid.free_set.opened);
+
+                grid.free_set.open(.{
+                    .encoded = free_set_checkpoint.buffer[0..free_set_checkpoint.size],
+                    .block_addresses = free_set_checkpoint.block_addresses[0..free_set_checkpoint.block_count()],
+                });
+                assert((free_set_checkpoint.size > 0) == (grid.free_set.count_acquired() > 0));
+                assert(grid.free_set.count_released() == grid.free_set_checkpoint.block_count());
+            }
+
+            grid.callback = .none;
+            callback(grid);
+        }
+
+        /// Checkpoint process is delicate:
+        ///   1. Encode free set.
+        ///   2. Derive the number of blocks required to store the encoding.
+        ///   3. Allocate free set blocks for the encoding (in the old checkpoint).
+        ///   4. Write the free set blocks to disk.
+        ///   5. Awaits all pending repair-writes to blocks that were just freed. This guarantees
+        ///      that there are no outstanding writes to (now-)free blocks when we enter the new
+        ///      checkpoint. This step runs concurrently to step 4.
+        ///   6. Mark currently released blocks as free and eligible for acquisition in the next
+        ///      checkpoint.
+        ///   7. Mark the free set's own blocks as released (but not yet free).
+        ///
+        /// This function handles step 1 and 5.
+        /// This function calls `free_set_checkpoint.checkpoint`, which handles steps 2-4.
+        /// The caller is responsible for calling FreeSet.checkpoint which handles 6 and 7.
         pub fn checkpoint(grid: *Grid, callback: *const fn (*Grid) void) void {
-            assert(grid.checkpointing == null);
-            assert(grid.canceling == null);
+            assert(grid.callback == .none);
```

Comment:
I was thinking about a pattern like

```
assert(grid.callback == .none);
defere assert(grid.callback == .checkpoint);
```

to underscore that this is a state transition function. But that's probably more noise than helpful.

Reply:
I do like that pattern when there are multiple branches / return points, though!